homeassistant/components/hyperion/light.py

```diff
@@ -78,27 +101,239 @@
+            data={
+                CONF_HOST: host,
+                CONF_PORT: port,
+            },
+        )
+        if (
+            result["type"] != data_entry_flow.RESULT_TYPE_CREATE_ENTRY
+            or result.get("result") is None
+        ):
+            _LOGGER.warning(
+                "Could not automatically migrate Hyperion YAML to a config entry."
+            )
+            return
+        config_entry = result.get("result")
+        result = await hass.config_entries.options.async_init(
+            config_entry.entry_id,
+            data={
+                CONF_PRIORITY: config[CONF_PRIORITY],
+            },
+        )
+        if result["type"] != data_entry_flow.RESULT_TYPE_CREATE_ENTRY:
+            _LOGGER.warning(
+                "Could not automatically migrate Hyperion YAML to an options entry."
+            )
+            return
+        _LOGGER.info(
+            "Successfully migrated Hyperion YAML configuration to a config entry."
+        )
+
+    # Kick off a config flow to create the config entry.
+    hass.async_create_task(
+        migrate_yaml_to_config_entry_and_options(host, port, config[CONF_PRIORITY])
+    )
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up a Hyperion platform from config entry."""
+    host = config_entry.data[CONF_HOST]
+    port = config_entry.data[CONF_PORT]
+    token = config_entry.data.get(CONF_TOKEN)
+
+    async def async_instances_to_entities(
+        platform: entity_platform.EntityPlatform,
+        response: Dict[str, Any],
+    ):
+        if not response or const.KEY_DATA not in response:
+            return
+        return await async_instances_to_entities_raw(
+            platform,
+            response[const.KEY_DATA],
+        )
+
+    async def async_instances_to_entities_raw(
+        platform: entity_platform.EntityPlatform,
+        instances: Dict[str, Any],
+    ):
+        registry = await async_get_registry(hass)
+        entities_to_add: List[Hyperion] = []
+        desired_instance_ids: Set[int] = set()
+
+        # In practice, an instance can be in 3 states as seen by this function:
+        #
+        #    * Exists, and is running: Add it to hass.
+        #    * Exists, but is not running: Cannot add yet, but should not delete it either. It will show up as "unavailable".
+        #    * No longer exists: Delete it from hass.
+
+        # Add instances that are missing (instance must be running)
+        for instance in instances:
+            instance_id = instance.get(const.KEY_INSTANCE)
+            if instance_id is None or not instance.get(const.KEY_RUNNING, False):
+                continue
+            desired_instance_ids.add(instance_id)
+            unique_id = get_hyperion_unique_id(server_id, instance_id)
+            entity_id = registry.async_get_entity_id(LIGHT_DOMAIN, DOMAIN, unique_id)
+            if entity_id is not None and entity_id in platform.entities:
```

Comment:
This check is not good. Users can disable entities in the entity registry and then they will not end up in `platform.entities`. If you want to prevent from adding something twice, instead of relying on HA data, keep a set around to track `(server_id, instance_id)` tuples.